### DISCUSSION TITLE
Renovate unnecessarily rebases PR

### DISCUSSION THREAD

**Discussion by oliviertassinari** [renovatebot/renovate#11083] [Request Help] [answered]
**How are you running Renovate?**

- [x] Self hosted

Please select which platform you are using:

- [x] github.com

**Describe the bug**

This is exactly as in #8981.

Once a maintainer merges one of Renovates PRs, it will rebase all its open PRs, although there should be no merge conflicts created by the recent merge.

This behavior is highly undesirable, since (in addition to taking a lot longer to merge PRs) it causes CI to run much more than necessary (increasing CI costs and consuming CI resources).
Any insight on what might be going on or how to fix it would be appreciated 🙏

**Relevant debug logs**

On the same branch:

- https://app.renovatebot.com/dashboard#github/mui-org/material-ui-x/415976551 rebase
- https://app.renovatebot.com/dashboard#github/mui-org/material-ui-x/416733053 rebase
- https://app.renovatebot.com/dashboard#github/mui-org/material-ui-x/416733130 no rebase
- https://app.renovatebot.com/dashboard#github/mui-org/material-ui-x/416734735 no rebase

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [x] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

We tried something to fix it in https://github.com/mui-org/material-ui/pull/27565 but it didn't help.

**Comment by oliviertassinari** [comment]
Oh actually, issue #8981 was moved to a discussion, https://github.com/renovatebot/renovate/discussions/8982. In our case, we have `"automerge": false,`, `rebaseWhen": "auto"`, and no "Require branches to be up to date before merging" on GitHub". So could it be a bug, where https://docs.renovatebot.com/updating-rebasing/#rebasing-out-of-date-branches is not respected?

**Comment by rarkins** [comment]
Do you have this setting checked in the repository settings?

![image](https://user-images.githubusercontent.com/6311784/128026385-9a89c08f-8739-4d3d-a8fb-5969bd6b9dff.png)


rebaseWhen=auto means branches are rebased if:
- They are conflicted, or
- You configured branches must be update to date AND the renovate branch fell behind the base branch

**Comment by oliviertassinari** [comment]
> Do you have this setting checked in the repository settings?

We don't 

<img width="458" alt="Capture d’écran 2021-08-03 à 15 48 09" src="https://user-images.githubusercontent.com/3165635/128026684-3503c5f5-6bd5-4fc9-9ad1-e434c1a8cf72.png">

**Comment by rarkins** [comment]
For the log messages you mentioned, can you locate one particular rebase which you're sure is incorrect, and copy/paste the relevant log messages for that branch back into this issue? The majority of people in this repo don't have access to view logs, even if open source as the app defaults to only allowing committers to the repo to view them.

**Comment by oliviertassinari** [comment, marked as answer]
> rebase which you're sure is incorrect

@rarkins Oh OK, I didn't know you don't have access to the logs.

Looking at the logs, there is something that is counter intuitive to me. The renovate build seems to compare the yarn.lock on a given PR between old and new rebased, it uses the equality to determine if it needs to rebase or not. To me it doesn't make sense. If I merge a change in the dependencies, the yarn.lock on HEAD is changing. The yarn.lock on the pull request is also going to change, so what? What I care about is conflits.

Could the issue be that renovate compare the yarn.lock regardless of if there is a conflit or not?

I'm adding the relevant log message with an actual reproduce once I get on a laptop '

**Reply by rarkins** [reply]
This is expected behaviour so I have converted it to a discussion.

What you are seeing is a byproduct of the combination of:
- Non-pinned dependencies
- "rangeStrategy": "update-lockfile"

In scenarios where you have ranges in `package.json` and rangeStrategy=update-lockfile, it means that changes are *only* to the `yarn.lock`. In such case Renovate will compute what the `yarn.lock` should be, and if it doesn't match what's already in the branch then it will push the new commit.

If instead of mostly had pinned dependencies then Renovate would instead base its rebase decision on whether the `package.json` needs to change or not. We may be able to adopt a similar approach one day for lock files too, but it will be a lot of work.

As a general rule though, it's risky not to rebase dependency update PRs if the base branch's lock file has changed (regardless of conflict status). Although git might happily merge the changed `yarn.lock`, it may no longer be valid or at least optimal for the changed `package.json`, so e.g. `yarn install --frozen-lockfile` could fail on your main branch after merging. There's also the possibility that the combination of changed dependencies which haven't been tested together could break your functionality/tests.

**Reply by oliviertassinari** [reply]
@rarkins Ok, thanks for the explanation. So we need "bump"?

**Reply by rarkins** [reply]
Sure, either pinning or bumping will avoid that raw lock file diff being done

**Reply by oliviertassinari** [reply]
It seems to work for us now with https://github.com/mui-org/material-ui/pull/27652
